backend/src/test/java/gov/cdc/usds/simplereport/api/BaseFullStackTest.java

```diff
@@ -0,0 +1,125 @@
+package gov.cdc.usds.simplereport.api;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import gov.cdc.usds.simplereport.config.authorization.UserPermission;
+import gov.cdc.usds.simplereport.db.model.ApiAuditEvent;
+import gov.cdc.usds.simplereport.db.model.auxiliary.HttpRequestDetails;
+import gov.cdc.usds.simplereport.service.AuditService;
+import gov.cdc.usds.simplereport.test_util.DbTruncator;
+import gov.cdc.usds.simplereport.test_util.TestDataFactory;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.BeforeEach;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
+
+/**
+ * Base class for all tests that simulate fully-integrated API requests by either patients or
+ * providers.
+ */
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
+public abstract class BaseFullStackTest {
+
+  @Autowired private DbTruncator _truncator;
+  @Autowired private AuditService _auditService;
+  @Autowired protected TestDataFactory _dataFactory;
+
+  protected Date _testStart;
+
+  @BeforeEach
+  void initTestStart() {
+    _testStart = new Date();
+  }
+
+  protected void truncateDb() {
+    _truncator.truncateAll();
+  }
+
+  protected ApiAuditEvent assertLastAuditEntry(
+      String username,
+      String operationName,
+      Set<UserPermission> permissions,
+      List<String> errorPaths) {
+    ApiAuditEvent event = getTimeCheckedEvent();
+    assertEquals(username, event.getUser().getLoginEmail());
+    assertEquals(operationName, event.getGraphqlQueryDetails().getOperationName());
+    if (permissions != null) {
+      assertEquals(
+          permissions.stream().map(UserPermission::name).collect(Collectors.toSet()),
+          Set.copyOf(event.getUserPermissions()),
+          "Recorded user permissions");
+    }
+    if (errorPaths == null) {
+      errorPaths = List.of();
+    }
+    assertEquals(errorPaths, event.getGraphqlErrorPaths(), "Query paths with errors");
+    return event;
+  }
+
+  /** REST audit event checker: includes "this is not graphql" checks" */
+  protected ApiAuditEvent assertLastAuditEntry(
+      HttpStatus status, String requestUri, String requestId) {
+    ApiAuditEvent event = getTimeCheckedEvent();
+    assertNull(event.getGraphqlQueryDetails());
+    assertEquals(null, event.getGraphqlErrorPaths());
```

Comment:
```suggestion
    assertNull(event.getGraphqlQueryDetails());
```

Reply:
Oops, I read this as `assertNull(X); assertEquals(null, X);` disregard

Reply:
I mean, the `assertEquals(null, Y)` still ought to be fixed, it's just a slightly different fix. 🙂